vagrant-tramp.el

```diff
@@ -13,70 +14,111 @@
+              (concat "_" name)))))
+
+(defun vagrant-tramp--running-boxes ()
+  "List as per `vagrant-tramp--all-boxes', but excluding boxes
+not reported to be running."
+  (-filter 'vagrant-tramp--box-running-p
+           (vagrant-tramp--all-boxes)))
+
+;;;###autoload
+(defun vagrant-tramp--completions (&optional file)
+  "List for vagrant tramp completion. FILE argument is ignored."
+  (--map (list nil it)
+         (-map 'vagrant-tramp--box-name
+               (vagrant-tramp--running-boxes))))
 
 ;;;###autoload
-(defun vagrant-tramp-term (box)
-  "SSH to a Vagrant BOX in an `ansi-term'."
+(defun vagrant-tramp-term (box-name)
+  "SSH into BOX using an `ansi-term'."
   (interactive
    (list
-    (let ((boxes (vagrant-tramp-list)))
-      (if (eq 1 (length boxes))
-          (car boxes)
-        (ido-completing-read "vagrant ssh to box: " boxes)))))
-
-  (let ((name (concat "vagrant ssh " box))
-        (cmd vagrant-tramp-ssh))
-    (with-current-buffer (make-term name cmd nil box)
-      (term-mode)
-      (term-char-mode)
-      (term-set-escape-char ?\C-x)
-      (switch-to-buffer (current-buffer)))))
+    (let* ((boxes (vagrant-tramp--running-boxes))
+           (names (-map 'vagrant-tramp--box-name boxes)))
+      (if (eq 1 (length names))
+          (car names)
+        (ido-completing-read "vagrant ssh to: " names)))))
+  (let* ((name (concat "vagrant terminal:" box-name))
+         (buffer (get-buffer-create (concat "*" name "*"))))
+    (unless (term-check-proc buffer)
+      (let* ((boxes (vagrant-tramp--running-boxes))
+             (box (--first (string=
+                            box-name
+                            (vagrant-tramp--box-name it))
+                           boxes))
+             (box-id (cdr (assoc 'id box))))
+        (with-current-buffer buffer
+          (cd (cdr (assoc 'dir box)))
+          (term-mode))
+        (term-exec buffer name "vagrant" nil (list "ssh" box-id))
+        (set-buffer buffer)
+        (term-mode)
+        (term-char-mode)))
+    (switch-to-buffer (concat "*" name "*"))))
 
 ;;;###autoload
-(defun vagrant-tramp-enable ()
+(defun vagrant-tramp-add-method ()
   "Add `vagrant-tramp-method' to `tramp-methods'."
   (add-to-list 'tramp-methods
                `(,vagrant-tramp-method
-                 (tramp-login-program     ,(shell-quote-argument vagrant-tramp-ssh))
+                 (tramp-login-program     ,vagrant-tramp-ssh)
                  (tramp-login-args        (("%h")))
                  (tramp-remote-shell      "/bin/sh")
-                 (tramp-remote-shell-args ("-c"))))
-  (tramp-set-completion-function vagrant-tramp-method
-                                 `((vagrant-tramp-parse ,vagrant-tramp-ssh))))
+                 (tramp-remote-shell-args ("-c")))))
+
+(defconst vagrant-tramp-completion-function-alist
+  '((vagrant-tramp--completions  ""))
+  "Default list of (FUNCTION FILE) pairs to complete vagrant method.")
+
+;;;###autoload
+(with-eval-after-load "tramp"
```

Comment:
You need a dependency on `(emacs "24.4")` in `Package-Requires:` for `with-eval-after-load`, or you need to stick to `eval-after-load` with a quoted `progn`.